### ISSUE TITLE
Wrong encoding

### ISSUE THREAD

**Issue by art-es** [antonioribeiro/countries#138]
How to fix the name of the country so that these strange characters do not appear?

In code:
![image](https://user-images.githubusercontent.com/31780910/71806413-cdca8880-3079-11ea-91f0-89326b92ff41.png)

In database:
![image](https://user-images.githubusercontent.com/31780910/71806476-f81c4600-3079-11ea-8dd0-fadb94d4338c.png)

I changed encoding table & tables to UTF-16LE:
![image](https://user-images.githubusercontent.com/31780910/71806548-1e41e600-307a-11ea-8aba-705832b72dbe.png)
![image](https://user-images.githubusercontent.com/31780910/71806569-2b5ed500-307a-11ea-92a8-944e3848f723.png)

**Comment by VictorPullzz** [comment]
I join this issue!

**Comment by remif25** [comment]
Same problem here. How can I fix it ?

**Comment by Marivint** [comment]
You can utf8_decode($countrie["name_en"])

**Comment by klodoma** [comment]
Can it be that the encoding in the backend is wrong? Or double encoded?

**Comment by devoncmather** [comment]
Same issue here

**Comment by klodoma** [comment]
> You can utf8_decode($countrie["name_en"])

Yes, this works, but it doesn't make sense to me. Seems that something is double encoded, haven't checked the sources though yet.

**Comment by giannicic** [comment]
I've made a check on the json countries file.

It seems that the double encoded values are the translated ones (Eg. the ones in the fields "name_XX").
For example Österreich is encoded in name_de as `"\u00c3\u0096sterreich"`
and an `utf8_decode` returns the correct value of `"\u00d6sterreich"`
which is the value under the `"name->native->bar->common"` field

**Comment by klodoma** [comment]
> For example Österreich is encoded in name_de as `"\u00c3\u0096sterreich"`
> and an `utf8_decode` returns the correct value of `"\u00d6sterreich"`

Yes, exactly. `utf8_decode` fixes it for the moment. 

We'll have to monitor when this gets fixed in the package, we'll have to remove our `utf8_decode` when this is done.

**Comment by ademtepe** [comment]
I have used the solution suggested here for Laravel Collection and it worked:

```
use PragmaRX\Countries\Package\Countries as Country;

Collection::macro('decode', function () {
    return $this->map(function ($value) {
        return utf8_decode($value);
    });
});

return Country::all()->pluck('name_tr', 'cca3')->decode();
```

**Comment by antonioribeiro** [comment]
The reason it's not done is because it's not easy to decode/re-encode them all correctly. Something I always have to say: the data we have here was not done by me, it's a collection of many other sources, and people just choose what they want/can use, I have zero control over this.

Unfortunately `utf8_decode();` is not a solution either. While trying to insert all cities into a PostgreSQL database I got this myself: 

<img width="1680" alt="Screenshot 2021-05-17 at 00 27 31" src="https://user-images.githubusercontent.com/3182864/118414747-ae4c6280-b6a6-11eb-8786-6cd736536200.png">

So if someone can come up with a strong solution for correctly enconding everything to UTF8, I'm more than pleased to merge a PR.

Cheers!

**Comment by antonioribeiro** [comment]
This is working for me:

``` php
protected function decode(?string $name): ?string
{
    if (blank($name) || mb_detect_encoding($name) !== 'UTF-8') {
        return $name;
    }

    return utf8_decode($name);
}
```

But I'm unsure if we have to do this in the package. I can't check if ALL encodings are good, and probably not every single will be fixed. Also, it will take a lot more time to generate all the files, which is already very slow. Any thoughts?

**Comment by antonioribeiro** [comment]
It didn't really fix them all, still had a lot of strings wrongly encoded, so I found [this forceutf8 package](https://github.com/neitanod/forceutf8) that solved it (not fully too, still got some wrong, but it's way better):

``` php
protected function decode(?string $name): ?string
{
    if (blank($name)) {
        return $name;
    }

    if (mb_detect_encoding($name) !== 'UTF-8') {
        return Encoding::toUTF8($name);
    }

    return Encoding::fixUTF8($name);
}

```

**Comment by klodoma** [comment]
@antonioribeiro from where are you getting the countries data or how are you putting that data together?

**Comment by antonioribeiro** [comment]
@klodoma , here you have a list of sources I'm using: https://github.com/antonioribeiro/countries#copyright. Sanitize encoding is not impossible, but it's a lot of data to sanitize. And they may require different strategies.

**Comment by lupinitylabs** [comment]
The issue is that part of the unicode is in unicode codepoint notation (`"common": "\u00d6sterreich"`), while a few lines down the same is encoded as a UTF-8 hex bytestring (`"name_de": "\u00c3\u0096sterreich"`). I can't imagine how the decoder should know what to do here. The first string is translated correctly into `Österreich`, while the second is translated into `\u00d6sterreich` (that's why the `utf8_decode` works for us in that case).

So, should we go with `utf8_decode`? Yes, but... be aware that if you are using one of the columns that encoded differently (like `name->common`), you will end up with a binary string representation:

```php
utf8_decode(json_decode('"'. "\u00d6sterreich" . '"'))
=> b"Österreich"
```

No fun... I would suggest rebuilding the whole JSON files with consistent encoding, or for that matter, I am going to go back to using [mledoze/countries](https://github.com/mledoze/countries), which was in better shape in that regard.

**Comment by ftrudeau-pelcro** [comment]
Indeed, encoding is inconsistent and I do believe @lupinitylabs's suggestion makes sense. Is there any solution brewing for this @antonioribeiro?
